src/game/requester.c

```diff
@@ -174,34 +189,28 @@ void Requester_SetHeading(REQUEST_INFO *req, const char *string)
     Text_Remove(req->heading);
     req->heading = NULL;
 
-    if (string) {
-        req->heading_text = string; // unsafe
-    } else {
-        req->heading_text = NULL;
-    }
+    size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+    snprintf(req->heading_text, out_size, "%s", string);
 }
 
 void Requester_ChangeItem(
     REQUEST_INFO *req, int32_t idx, const char *string, uint16_t flag)
 {
-    if (string) {
-        strcpy(&req->item_texts[idx * req->item_text_len], string);
+    if (idx > 0 && idx < req->max_items && string) {
+        size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+        snprintf(req->item_texts[idx], out_size, "%s", string);
         req->item_flags[idx] = flag;
-    } else {
-        req->item_flags[idx] = 0;
     }
 }
 
 void Requester_AddItem(REQUEST_INFO *req, const char *string, uint16_t flag)
 {
-    if (string) {
-        strcpy(&req->item_texts[req->items * req->item_text_len], string);
-        req->item_flags[req->items] = flag;
-    } else {
-        req->item_flags[req->items] = 0;
+    if (req->items_used < req->max_items && string) {
+        size_t out_size = snprintf(NULL, 0, "%s", string) + 1;
+        snprintf(req->item_texts[req->items_used], out_size, "%s", string);
+        req->item_flags[req->items_used] = flag;
+        req->items_used++;
```

Comment:
Don't we need to zero the item flags and the text buffer in case the string is NULL?

Reply:
`if (req->items_used < req->max_items && string) {`

`Requester_AddItem` just doesn't run at all `string` is `NULL`. Is that ok or do you want to change it?

Reply:
Sorry - this makes sense. I don't think we'll ever try to add a NULL item anyway.